app/forms/hyrax/forms/permission_template_form.rb

```diff
@@ -38,14 +38,61 @@ def initialize(model)
         select_release_varies_option(model)
       end
 
+      # @return [Hash] { :content_tab (for confirmation message) , :updated (true/false), :error_code (for flash error lookup) }
       def update(attributes)
-        update_admin_set(attributes)
-        update_permission_template(attributes)
-        grant_workflow_roles(attributes)
+        return_info = { content_tab: tab_to_update(attributes) }
+        update_status = {}
+        case return_info[:content_tab]
+        when "participants"
+          update_status = update_participants_options(attributes)
+        when "visibility"
+          update_status = update_visibility_options(attributes)
+        when "workflow"
+          update_status = update_workflow_options(attributes)
+        end
+        return_info.merge!(update_status)
       end
 
       private
 
+        # @return [String]
+        def tab_to_update(attributes)
+          return "participants" if attributes[:access_grants_attributes].present?
+          return "workflow" if attributes[:workflow_id].present?
+          return "visibility" if attributes.key?(:visibility)
+        end
+
+        # @return [Hash] { :updated } = true
+        def update_participants_options(attributes)
+          return_info = {}
+          update_admin_set(attributes)
+          update_permission_template(attributes)
+          return_info[:updated] = true
+          return_info
+        end
+
+        # @return [Hash] { :error_code - used for flash notice, :updated - true or false}
+        def update_visibility_options(attributes)
+          return_info = {}
+          validated_attributes = validate_visibility_combinations(attributes)
+          if validated_attributes[:valid] == true
+            update_permission_template(attributes)
+          else
+            return_info[:error_code] = validated_attributes[:error_code]
+          end
+          return_info[:updated] = validated_attributes[:valid]
+          return_info
+        end
+
+        # @return [Hash] { :updated } = true
```

Comment:
Seems like an odd return value, since it is never informative to the caller.  Might as well return null or whatever `grant_workflow_roles` returns, right?  The `{}` are not conventional in a YARD description, but that aside, I think you want `true` inside them (matching to the other `@return` statements you have).